extmod/ujson: Add support for CPython-compatible nan/inf handling.

stinos opened the pull request:
Modify the JSON loading code so it can parse the JavaScript floating
point representations NaN/Infinity/-Infinity, and support the allow_nan
keyword argument for dump() and dumps() which controls whether those
representations get created or else lead to raising a ValueError.
All of this is the same as CPython's implementation and when enabled
fixes the inconsistency in the current code which unconditionally
creates lowercase nan/inf strings but doesn't parse them.

This is a fix for #3511. It requires more changes than I initially anticipated, but I don't see another way than to pass the needed information into `float_print` via the `mp_print_kind` argument.
Also needs quite some extra code, enabled by default for a couple of ports but that was mainly to get CI feedback.
Flag is called MICROPY_PY_UJSON_ALLOWNAN but it's essentially about CPython compatibility so another name might be more suitable.

stinos commented:
Alternative more micro fix would be to drop the `allow_nan` keyword and instead make the behaviour fixed but selectable at built time. I'm thinking
```
MICROPY_PY_UJSON_SPEC = 0 -> current behaviour
MICROPY_PY_UJSON_SPEC = 1 -> fully compliant, no out of range numbers supported
                             (with rasing of exceptions both in dump and load)
MICROPY_PY_UJSON_SPEC = 2 -> JavaScript extension style, out of range numbers supported
```

dpgeorge commented:
Thanks for posting this.  It is quite a large patch (mostly in the loading) for what is an extension to JSON.  So not really sure if it's something we should add.  Reading through #3511 the main point there was that it's unexpected that dump(s) puts out nan/inf, so the simplest fix for that would be to raise an exception if trying to dump these values.  That would be your option "1" above.

We can let this sit a bit for other feedback.

dpgeorge commented:
@stinos any update on this PR, are you using it, did you find a better solution?

stinos commented:
Didn't work on this any further, but have been using it for years. I can have a look again, but which direction should I be looking in? Try to simplify this implementation, or look into raising exceptions or something like in the second comment?

dpgeorge commented:
> I can have a look again, but which direction should I be looking in?

I'm not sure.... I just wanted to see if there was anything that would sway this PR one way or the other.  JSON is used a lot with uPy and there are continuous requests to improve it (see also #7478), so probably makes sense to include this.

stinos committed:
py: Extract floating point routines from objfloat.c.

Especially the fp_is* functions are useful to have in the public
API, and will be used in upcoming commits regarding JSON.

stinos committed:
extmod/ujson: Support CPython-compatible nan/inf handling.

stinos committed:
tests/extmod/ujson: Add tests for special float handling.

stinos committed:
extmod/ujson: Be more strict when dumping special float numbers.

Add an option to raise an exception instead of creating invalid
JSON text.  Implemented with a 'mode' preprocessor definition so
other behavior can be added later on.
